File: src/bumperbot_py_examples/bumperbot_py_examples/simple_turtlesim_kinematics.py
#!/usr/bin/env python3

import rclpy
import numpy as np
from rclpy.node import Node
from turtlesim.msg import Pose

class SimpleTurtlesimKinematics(Node):
    
    def __init__(self):
        super().__init__('simple_turtlesim_kinematics')
        
        self.turtle1_pose_sub_ = self.create_subscription(
            Pose,
            '/turtle1/pose',
            self.turtle1PoseCallback,
            10
        )

        self.turtle2_pose_sub_ = self.create_subscription(
            Pose,
            '/turtle2/pose',
            self.turtle2PoseCallback,
            10
        )
        
        self.last_turtle1_pose_ = Pose()
        self.last_turtle2_pose_ = Pose()

    def turtle1PoseCallback(self, msg):
        self.last_turtle1_pose_ = msg
        
    def turtle2PoseCallback(self, msg):
        self.last_turtle2_pose_ = msg

        Tx = self.last_turtle2_pose_.x - self.last_turtle1_pose_.x
        Ty = self.last_turtle2_pose_.y - self.last_turtle1_pose_.y

        theta = self.last_turtle2_pose_.theta - self.last_turtle1_pose_.theta
        theta_deg = 180*theta/np.pi
    
        self.get_logger().info(f"""\n
            Translation vector turtle1 -> turtle2 \n
            Tx: {Tx:.2f} \n
            Ty: {Ty:.2f} \n
            Rotation matrix turtle1 -> turtle2 \n
            theta (rad): {theta:.2f} rad \n
            theta (deg): {theta_deg:.2f} deg \n
            [R11 R12] :    [{np.cos(theta):.2f}\t{-np.sin(theta):.2f}] \n
            [R21 R22] :    [{np.sin(theta):.2f}\t{np.cos(theta):.2f}] \n
            """) 

def main():
    rclpy.init()
    simple_turtlesim_kinematics = SimpleTurtlesimKinematics()
    rclpy.spin(simple_turtlesim_kinematics)
    simple_turtlesim_kinematics.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_turtlesim_kinematics.cpp
#include "bumperbot_cpp_examples/simple_turtlesim_kinematics.hpp"
#include <rclcpp/rclcpp.hpp>
#include <turtlesim/msg/pose.hpp>

// constructor, initialize Node
SimpleTurtlesimKinematics::SimpleTurtlesimKinematics(
    const std::string &node_name) : Node(node_name)
{
    // Create a subscriber on the "/turtle1/pose" topic
    turtle1_pose_sub_ = create_subscription<turtlesim::msg::Pose>(
        "/turtle1/pose", // topic name
        10,              // size of buffer queue
        std::bind(&SimpleTurtlesimKinematics::turtle1PoseCallback,
                  this,
                  std::placeholders::_1) // callback function
    );

    RCLCPP_INFO(get_logger(), "Listening in topic /turtle1/pose");

    // Create a subscriber on the "/turtle2/pose" topic
    turtle2_pose_sub_ = create_subscription<turtlesim::msg::Pose>(
        "/turtle2/pose", // topic name
        10,              // size of buffer queue
        std::bind(&SimpleTurtlesimKinematics::turtle2PoseCallback,
                  this, std::placeholders::_1) // callback function
    );

    RCLCPP_INFO(get_logger(), "Listening in topic /turtle2/pose");
}

void SimpleTurtlesimKinematics::turtle1PoseCallback(const turtlesim::msg::Pose::SharedPtr pose)
{
    last_turtle1_pose_ = pose;
}

void SimpleTurtlesimKinematics::turtle2PoseCallback(const turtlesim::msg::Pose::SharedPtr pose)
{
    last_turtle2_pose_ = pose;
    float Tx = last_turtle2_pose_->x - last_turtle1_pose_->x;
    float Ty = last_turtle2_pose_->y - last_turtle1_pose_->y;
    float theta = last_turtle2_pose_->theta - last_turtle1_pose_->theta;
    float theta_deg = theta * 180 / M_PI;

    RCLCPP_INFO_STREAM(get_logger(),
                       "\nTranslation vector turtle1 -> turtle2\n"
                           << "Tx: " << Tx << "\n"
                           << "Ty: " << Ty << "\n"
                           << "Rotation matrix turtle1 -> turtle2\n" 
                           << "Theta(rad): " << theta << " rad\n"
                           << "Theta(deg): " << theta_deg << " deg\n"
                           << "[R11 R12] :    [" << std::cos(theta) << "\t" << -std::sin(theta) << "] \n"
                           << "[R21 R22] :    [" << std::sin(theta) << "\t" << std::cos(theta) << "] \n"
                           );
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv); // initialize ROS2

    auto node = std::make_shared<SimpleTurtlesimKinematics>("simple_kinematics_node"); // create a shared pointer to the SimpleTurtlesimKinematics class

    rclcpp::spin(node); // run the node

    rclcpp::shutdown(); // shutdown ROS2

    return 0;
}